src/utils/mtev_lru.c

```diff
@@ -0,0 +1,252 @@
+
+static void
+hs_init(ck_hs_t *hs, unsigned int mode, ck_hs_hash_cb_t *hf, ck_hs_compare_cb_t *cf, unsigned long size)
+{
+  if(!rand_init) {
+    srand48((long int)time(NULL));
+    rand_init = 1;
+  }
+  struct ck_malloc *allocator = &malloc_ck_hs;
+
+  if (ck_hs_init(hs, mode | CK_HS_MODE_SPMC, hf, cf, allocator, size, lrand48()) == false) {
+    mtevFatal(mtev_error, "Cannot initialize ck_hs\n");
+  }
+}
+
+static inline bool
+hs_string_compare(const void *a, const void *b)
+{
+  return strcmp((const char * const)a, (const char * const)b) == 0;
+}
+
+mtev_lru_t *
+mtev_lru_create(int32_t max_entries, void (*free_fn)(void *)) 
+{
+  struct mtev_lru *r = malloc(sizeof(struct mtev_lru));
+  if (max_entries <= 0) {
+    hs_init(&r->hash, CK_HS_MODE_OBJECT | CK_HS_MODE_DELETE, lru_entry_hash, hs_string_compare, 1024);
+  } else {
+    hs_init(&r->hash, CK_HS_MODE_OBJECT | CK_HS_MODE_DELETE, lru_entry_hash, hs_string_compare, max_entries * 2);
+  }
+  TAILQ_INIT(&r->lru_cache);
+  r->lru_cache_size = 0;
+  r->expire_count = 0;
+  r->max_entries = max_entries;
+  if (free_fn != NULL) {
+    r->free_fn = free_fn;
+  } else {
+    r->free_fn = free;
+  }
+  pthread_mutex_init(&r->mutex, NULL);
+  return r;
+}
+
+void
+mtev_lru_destroy(mtev_lru_t *lru) 
+{
+  mtev_lru_invalidate(lru);
+  ck_hs_destroy(&lru->hash);
+  free(lru);
+}
+
+void
+mtev_lru_invalidate(mtev_lru_t *lru)
+{
+  pthread_mutex_lock(&lru->mutex);
+  void *entry;
+  ck_hs_iterator_t it = CK_HS_ITERATOR_INITIALIZER;
+  while(ck_hs_next(&lru->hash, &it, &entry) == true) {
+    struct lru_entry *e = container_of(entry, struct lru_entry, key);
+    lru->free_fn(e->entry);
+    free(e);
+  }
+  lru->lru_cache_size = 0;
+  pthread_mutex_unlock(&lru->mutex);
+}
+
+#define GC_CADENCE 10000
+
+/* this requires a lock to be held elsewhere */
+static void
+expire_oldest_lru_cache_no_lock(mtev_lru_t *c)
+{
+  /* remove from TAILQ */
+  struct lru_entry *e = TAILQ_LAST(&c->lru_cache, lru_list);
+  TAILQ_REMOVE(&c->lru_cache, e, list_entry);
+  c->lru_cache_size--;
+
+  /* remove from hs */
+  unsigned long hash = CK_HS_HASH(&c->hash, lru_entry_hash, e->key);
+  ck_hs_remove(&c->hash, hash, e->key);
+
+  int ec = ck_pr_load_32(&c->expire_count);
+  if (ec == GC_CADENCE) {
+    c->expire_count = 0;
+    ck_hs_gc(&c->hash, GC_CADENCE, (rand() % ck_hs_count(&c->hash)));
+  }
+  c->expire_count++;
+
+  if (e->entry != NULL) {
+    c->free_fn(e->entry);
+  }
+  free(e);
+}
+
+static void 
+add_lru_cache_no_lock(mtev_lru_t *c, struct lru_entry *e)
+{
+  if (c->max_entries > 0) {
+    if ((c->lru_cache_size + 1) > c->max_entries) {
+      expire_oldest_lru_cache_no_lock(c);
+    }
+    TAILQ_INSERT_HEAD(&c->lru_cache, e, list_entry);
+  }
+  c->lru_cache_size++;
+}
+
+static void 
+touch_lru_cache_no_lock(mtev_lru_t *c, struct lru_entry *e)
+{
+  if (c->max_entries > 0) {
+    if (e->list_entry.tqe_next != NULL || e->list_entry.tqe_prev != NULL) {
```

Comment:
I assume this function relies on the LRU entry being on the lru_cache, in which case you don't need to move the entry if it's already at the head. And it's probably better to use the TAILQ macros to peek into the structure, than to rely on their implementation details?
```
if (TAILQ_PREV(e, list_entry) != NULL) {
  TAILQ_REMOVE(&c->lru_cache, e, list_entry);
  TAILQ_INSERT_HEAD(&c->lru_cache, e, list_entry);
}
```